### PR TITLE
Initial autocompletion

### PR DESCRIPTION
**Pull request by bthesorceror**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -18,7 +18,8 @@ path = "src/lib.rs"
 serde_json = "1.0"
 serde_repr = "0.1"
 serde = { version = "1.0.101", features = ["derive"] }
-flux = { git = "https://github.com/influxdata/flux", rev = "8e956e67" }
+flux = { git = "https://github.com/influxdata/flux", rev = "1c7cc9f0" }
+libstd = { git = "https://github.com/influxdata/flux", rev = "1c7cc9f0" }
```
Comment (jademcgough) [inline Cargo.toml]: isn't this just reimporting flux as another name?

Comment (jademcgough) [reply Cargo.toml]: nevermind just found https://github.com/influxdata/flux/blob/8fd3eae64365b839ed1c1477842cf56b745470fb/libflux/src/libstd/Cargo.toml :)

Comment (bthesorceror) [reply Cargo.toml]: no, there are two libraries in that repo and we have to explicitly get both.

File: src/protocol/properties.rs [hunk 2]
```diff
@@ -209,7 +209,15 @@ pub enum TextDocumentSyncKind {
     Incremental = 2,
 }
 
-#[derive(Serialize, Deserialize, Clone)]
+#[derive(Serialize, Clone)]
+pub struct CompletionOptions {
+    #[serde(rename = "resolveProvider")]
+    pub resolve_provider: Option<bool>,
+    // #[serde(rename = "triggerCharacters")]
```
Comment (jademcgough) [inline src/protocol/properties.rs]: Can this commented out bit be removed?

Comment (bthesorceror) [reply src/protocol/properties.rs]: sure!

File: src/handler.rs [hunk 3]
```diff
@@ -83,6 +85,11 @@ impl Handler {
             Box::new(DocumentSymbolHandler::default()),
         );
 
+        mapping.insert(
+            "textDocument/completion".to_string(),
```
Comment (jademcgough) [inline src/handler.rs]: Could we add a new handler test for this?

Comment (bthesorceror) [reply src/handler.rs]: yeah, I was just holding off till I could get some feed back that everything is the way we want it to work.

File: src/handlers/completion.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    position: Position,
+) -> Result<Option<String>, String> {
+    let pkg = utils::create_semantic_package(uri.clone())?;
+    let walker = Rc::new(walk::Node::Package(&pkg));
+    let mut visitor = NodeFinderVisitor::new(position);
+
+    walk::walk(&mut visitor, walker);
+
+    let state = visitor.state.borrow();
+    let node = (*state).node.clone();
```
Comment (kelwang) [inline src/handlers/completion.rs]: you can just borrow here

File: src/handlers/completion.rs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+    if let Some(node) = node {
+        match node.as_ref() {
+            Node::Identifier(ident) => {
+                let name = ident.name.clone();
+                return Ok(Some(name));
+            }
+            Node::IdentifierExpr(ident) => {
+                let name = ident.name.clone();
```
Comment (kelwang) [inline src/handlers/completion.rs]: borrow here &ident.name

File: src/handlers/completion.rs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+    walk::walk(&mut visitor, walker);
+
+    let state = visitor.state.borrow();
+    let node = (*state).node.clone();
+
+    if let Some(node) = node {
+        match node.as_ref() {
+            Node::Identifier(ident) => {
+                let name = ident.name.clone();
```
Comment (kelwang) [inline src/handlers/completion.rs]: &ident.name

**Pull request merged**